### PR TITLE
Add constraints, use them for `test_zero` and friends

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR
* adds an `add_constraint` helper
* adds assertions `assert_is_zero`, `assert_equal`, and `assert_boolean`
* adds `is_zero` and `equal` helper functions
* replaces existing unsafe uses of `test_zero` with `is_zero` or `equal`.

### CODE REVIEW

File: optimism/src/mips/interpreter.rs [hunk 1]
```diff
@@ -572,6 +603,41 @@ pub trait InterpreterEnv {
...
+    /// The value returned may be a placeholder; callers should be careful not to depend directly
+    /// on the value stored in the variable.
+    unsafe fn inverse_or_zero(
+        &mut self,
+        x: &Self::Variable,
+        position: Self::Position,
+    ) -> Self::Variable;
+
+    fn is_zero(&mut self, x: &Self::Variable) -> Self::Variable {
+        let res = {
```
Comment (querolita) [inline optimism/src/mips/interpreter.rs:622]: maybe call this value `is_zero`

File: optimism/src/mips/interpreter.rs [hunk 2]
```diff
@@ -572,6 +603,41 @@ pub trait InterpreterEnv {
...
+            let pos = self.alloc_scratch();
+            unsafe { self.test_zero(x, pos) }
+        };
+        let x_inv_or_zero = {
+            let pos = self.alloc_scratch();
+            unsafe { self.inverse_or_zero(x, pos) }
+        };
+        // If x = 0, then res = 1 and x_inv_or_zero = _
+        // If x <> 0, then res = 0 and x_inv_or_zero = x^(-1)
+        self.add_constraint(x.clone() * x_inv_or_zero.clone() + res.clone() - Self::constant(1));
```
Comment (querolita) [inline optimism/src/mips/interpreter.rs:632]: looks good to me

File: optimism/src/mips/interpreter.rs [hunk 3]
```diff
@@ -802,16 +868,8 @@ pub fn interpret_rtype<Env: InterpreterEnv>(env: &mut Env, instr: RTypeInstructi
...
-                // FIXME: Requires constraints
-                let pos = env.alloc_scratch();
-                unsafe { env.test_zero(&(syscall_num.clone() - Env::constant(SYSCALL_BRK)), pos) }
-            };
-            let is_sysclone = {
-                // FIXME: Requires constraints
-                let pos = env.alloc_scratch();
-                unsafe { env.test_zero(&(syscall_num.clone() - Env::constant(SYSCALL_CLONE)), pos) }
-            };
+            let is_sysbrk = env.equal(&syscall_num, &Env::constant(SYSCALL_BRK));
```
Comment (querolita) [inline optimism/src/mips/interpreter.rs:871]: looks lovely

**Review by querolita — APPROVED** [review]
Lovely

**Review by dannywillems — APPROVED** [review]
Bumping @querolita approval.

**Pull request merged**